mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,281 @@
+                )
+            elif self.service_type == ServiceTypeEnum.load_balancer:
+                host, port = LoadBalancerService().get_host_and_port(service)
+            elif self.service_type == ServiceTypeEnum.cluster_ip:
+                raise TypeError(
+                    "Cannot expose service of type ClusterIP outside the Kubernetes Cluster"
+                )
+            else:
+                raise ValueError(
+                    f"service_type supplied is {self.service_type}, valid values are {[e.value for e in ServiceTypeEnum]} only"
+                )
+        except (IndexError, ValueError, TypeError) as e:
+            print("Couldn't determine host and port from the service deployed")
+            raise e
+        if host is not None and port is not None:
+            return HTTPClient(host=host, port=port)
+        raise TypeError(
+            f"host and port determined are not valid, received host as {host} and port as {port}"
+        )
+
+
+class K8sEnv(MlemEnv[K8sDeployment]):
+    """MlemEnv implementation for Kubernetes Environments"""
+
+    type: ClassVar = "kubernetes"
+    deploy_type: ClassVar = K8sDeployment
+    """type of deployment being used for the Kubernetes environment"""
+
+    registry: Optional[DockerRegistry] = None
+    """docker registry"""
+    templates_dir: List[str] = []
+    """list of dirs where templates reside"""
+
+    def get_registry(self, meta: K8sDeployment):
+        registry = meta.registry or self.registry
+        if not registry:
+            raise MlemError(
+                "registry to be used by Docker is not set or supplied"
+            )
+        return registry
+
+    def get_image_name(self, meta: K8sDeployment):
+        return meta.image_name or generate_docker_container_name()
+
+    def get_server(self, meta: K8sDeployment):
+        return (
+            meta.server
+            or project_config(
+                meta.loc.project if meta.is_saved else None
+            ).server
+        )
+
+    def deploy(self, meta: K8sDeployment):
+        self.check_type(meta)
+        redeploy = False
+        with meta.lock_state():
+            meta.load_kube_config()
+            state: K8sDeploymentState = meta.get_state()
+            if state.image is None or meta.model_changed():
+                image_name = self.get_image_name(meta)
+                state.image = build_k8s_docker(
+                    meta=meta.get_model(),
+                    image_name=image_name,
+                    registry=self.get_registry(meta),
+                    daemon=meta.daemon,
+                    server=self.get_server(meta),
+                )
+                meta.update_model_hash(state=state)
+                redeploy = True
+
+            if state.deployment_name is None or redeploy:
+                generator = K8sYamlGenerator(
+                    namespace=meta.namespace,
+                    image_name=state.image.name,  # type: ignore
```

Comment:
why # type: ignore?

Reply:
because `state.image` is `Optional[DockerImage]` which complains

```
Item "None" of "Optional[DockerImage]" has no attribute "name"
```

Reply:
That means you should check if it not `None`, not just ignore the warning

Reply:
fixed